RawSpeed/Camera.cpp

```diff
@@ -27,115 +28,102 @@ namespace RawSpeed {
+  if (make.empty())
     ThrowCME("Camera XML Parser: \"make\" attribute not found.");
-  make = canonical_make = key.as_string();
 
-  key = camera.attribute("model");
-  if (!key)
+  model = canonical_model = canonical_alias = camera.attribute("model").as_string();
+  // chdk cameras seem to have an empty model?
+  if (!camera.attribute("model")) // (model.empty())
     ThrowCME("Camera XML Parser: \"model\" attribute not found.");
-  model = canonical_model = canonical_alias = key.as_string();
 
   canonical_id = make + " " + model;
 
-  supported = true;
-  key = camera.attribute("supported");
-  if (key) {
-    string s = string(key.as_string());
-    if (s.compare("no") == 0)
-      supported = false;
-  }
+  supported = camera.attribute("supported").as_string("yes") == string("yes");
+  mode = camera.attribute("mode").as_string("");
+  decoderVersion = camera.attribute("decoder_version").as_int(0);
 
-  key = camera.attribute("mode");
-  if (key) {
-    mode = key.as_string();
-  } else {
-    mode = string("");
-  }
-
-  key = camera.attribute("decoder_version");
-  if (key) {
-    decoderVersion = key.as_int(0);
-  } else {
-    decoderVersion = 0;
-  }
-
-  for (xml_node node = camera.first_child(); node; node = node.next_sibling()) {
-    parseCameraChild(node);
+  for (xml_node c : camera.children()) {
+    parseCameraChild(c);
   }
 }
 
-Camera::Camera( const Camera* camera, uint32 alias_num) : cfa(iPoint2D(0,0))
+Camera::Camera(const Camera* camera, uint32 alias_num) : cfa(iPoint2D(0,0))
 {
   if (alias_num >= camera->aliases.size())
     ThrowCME("Camera: Internal error, alias number out of range specified.");
 
-  make = camera->make;
+  *this = *camera;
   model = camera->aliases[alias_num];
-  canonical_make = camera->canonical_make;
-  canonical_model = camera->canonical_model;
   canonical_alias = camera->canonical_aliases[alias_num];
-  canonical_id = camera->canonical_id;
-  mode = camera->mode;
-  cfa = camera->cfa;
-  supported = camera->supported;
-  cropSize = camera->cropSize;
-  cropPos = camera->cropPos;
-  decoderVersion = camera->decoderVersion;
-  for (uint32 i = 0; i < camera->blackAreas.size(); i++) {
-    blackAreas.push_back(camera->blackAreas[i]);
-  }
-  for (uint32 i = 0; i < camera->sensorInfo.size(); i++) {
-    sensorInfo.push_back(camera->sensorInfo[i]);
-  }
-  map<string,string>::const_iterator mi = camera->hints.begin();
-  for (; mi != camera->hints.end(); ++mi) {
-    hints.insert(make_pair((*mi).first, (*mi).second));
-  }
+  aliases.clear();
+  canonical_aliases.clear();
 }
 
-Camera::~Camera(void) {
+static string name(const xml_node &a) {
+  return string(a.name());
 }
 
-static bool isTag(const char_t *a, const char* b) {
-  return 0 == strcmp(a, b);
-}
-
-void Camera::parseCameraChild(xml_node &cur) {
-  if (isTag(cur.name(), "CFA")) {
-    if (2 != cur.attribute("width").as_int(0) || 2 != cur.attribute("height").as_int(0)) {
-      supported = FALSE;
```

Comment:
This check went to Nirvana :) (and on purpose) because I deemed it to be useless for 3 reasons:
1. it does not make sure that the given data in `CFA` is consistent, e.g. there could be single entries missing or mentioned twice
1. there is no equivalent check in the `CFA2` case
1. it is actually no problem at all to specify all of the info from a `CFA2` tag inside a `CFA` tag, so widths other than 2 are generally no problem and have been supported by the code

In fact I deem this CFA/CFA2 situation another opportunity for improvement but it was very low on my priority list. I'll open an issue for more focused discussion.